CONFIGURATION.md

```diff
@@ -0,0 +1,201 @@
+- detailed and generated documentation for all configuration properties
+
+## Proposal: We keep YAML as the configuration format
+
+While I have a good collection of scars from dealing with YAML (especially in CircleCI), I think the format is moderately more readable than JSON and almost as widely accepted, which makes it worthwhile suffering through it instead of picking something that will feel uncommon or out of the time.
+
+## Proposal: We introduce a JSON Schema for configuration files
+
+This way you can see errors in configuration files while editing them, and we have a formal way of describing and documenting the configuration format.
+
+## Proposal: Strains are the only top-level concept
+
+This means that there are no concepts that are orthogonal to strains, but that every other configuration object can be assigned to one or more strains, and that every request can be assigned to one specific strain.
+
+There will be no reconciliation of strains and some other concept at runtime. Every .hlx/strains.json will contain a fully deterministic, fully resolved and denormalized representation of all strains.
+
+## Proposal: Other complex configuration objects can be defined in a definitions container and re-used
+
+In order to make writing configurations easier and to enable re-use of configuration settings, we introduce a definitions container at the top level of the `helix-config.yaml` that can hold re-usable definitions like this:
+
+```yaml
+definitions:
+  strains:
+    base: &basestrain
+      code: /acapt/default/https---github-com-adobe-project-helix-io-git--master--
+      sticky: false
+    
+  origins:
+    publish: &publish
+      address: 192.168.0.1
+      name: publish
+      use_ssl: false
+```
+
+## Proposal: Configuration objects can be referenced using YAML references
+
+In order to re-use a complex configuration object (like an origin, a mount mapping, or even a strain), use YAML references (`*publish`) or YAML extensions (`<< *basestrain`). This allows us to rely on the behavior of the YAML parser instead of using home-grown inheritance and referencing logic that is likely less robust and certainly less well documented.
+
+A strain re-using the `*basestrain` above would look like this:
+
+```yaml
+  client:
+    <<: *basestrain
+    condition: req.http.host == "client.project-helix.io"
+    content:
+      repo: helix-cli
+      ref: master
+      owner: adobe
+    directoryIndex: readme.html
+```
+
+A strain referencing the `*publish` origin would look like this:
+```yaml
+  proxy-detailed:
+    sticky: true
+    origin: *publish
+```
+
+## Proposal: The `default` strain serves all traffic in production when no other strain is selected, therefore a `default` strain is mandatory
+
+We need one strain that is serving traffic that no other strain claims. This could be one by:
+
+1. requiring the existence of a `default` strain
+2. requiring the existence of one and only one strain that has the `default: true` property set
```

Comment:
3. have a toplevel configuration value: `defaultStrain` that points to the default strain.

Reply:
(3) would be even harder to validate.